### PR TITLE
Numpy bfloat16 workaround

### PR DESCRIPTION
**Pull request by hunhoffe**
This PR conditionally installs `bfloat16` where possible and ignores errors otherwise. This is to allow some `bfloat16` integration in the mlir-aie python code. A workaround is necessary because the actual `bfloat16` library does not seem to compile well in vscode environments, such as used in the CI. While another long-term solution is likely necessary, this workaround allows one to create designs that look as though `bfloat16` is working properly.

The conditional installation looks like this in setup scripts and CI actions:
```bash
 pip install -r python/requirements_bfloat16.txt || echo "Failed to install bfloat16; this is ok!"
```

The conditional import in the python looks like this:
```python
class _pseudo_bfloat16(np.float16):
    def __init__(*args, **kwargs):
        raise TypeError(
            "This is a placeholder class for bfloat16. Install bfloat16 python package in order to instantiate bfloat16 values"
        )

try:
    from bfloat16 import bfloat16 as real_bfloat16

    bfloat16 = real_bfloat16
except:
    # Numpy doesn't support bfloat16 at this time. Using hacky extension of np.float16 as a placeholder, for now.
    print(
        "Warning! bfloat16 python package not available, so using placeholder bfloat16 type",
        file=sys.stderr,
    )
    bfloat16 = _pseudo_bfloat16
```

And finally, programming examples can import a version of `bfloat16` (either real or placeholder) like so:
```python
from aie.extras.util import bfloat16
```

This PR also diasables some of the AIE distros github action, to disable MacOS builds/tests, as those aren't currently supported. They didn't work before this PR, however, the action only runs when you change the workflow definition, so it was hidden from most PRs/CI runs.

### CODE REVIEW

**Comment by hunhoffe** [comment]
@andrej tagging because this changes the matmul examples!

File: programming_examples/basic/matrix_multiplication/single_core/aie2.py [hunk 1]
```diff
@@ -73,24 +82,8 @@ def my_matmul(M, K, N, m, k, n, dtype_in_str, dtype_out_str):
...
-    elif dtype_out_str == "i8":
-        dtype_out = T.i8
-    elif dtype_out_str == "i16":
-        dtype_out = T.i16
-    elif dtype_out_str == "f32":
-        dtype_out = T.f32
-    elif dtype_out_str == "i32":
-        dtype_out = T.i32
+    dtype_in = dtype_map[dtype_in_str]
+    dtype_out = dtype_map[dtype_out_str]
```
Comment (andrej) [inline programming_examples/basic/matrix_multiplication/single_core/aie2.py:86]: much better! thanks for cleaning up my mess!

**Review by andrej — APPROVED** [review]
Thanks for tagging me. I only looked at the matmul changes. They look good, but please keep in mind that the CI currently only tests a limited number of data types. You can see which ones the CI tests in the `run_makefile*.lit` files. (If the `.lit` does not specify, the test will use the default data type in the `makefile-common`.)

Since the matmul tests are already some of the slowest in the CI, and the changes seem very symmetrical for all data types, it probably doesn't make sense to add a bunch more CI tests. But I would advise to test everything with all data types at least locally! Especially the bf16 one.

**Comment by hunhoffe** [comment]
@andrej Thanks for taking a look! I did test locally (only for default shape) through changing the input datatype in `makefile-common` to bf16

**Review by fifield — APPROVED** [review]

**Pull request merged**